File: myfidpass/Services/NotificationSendLocalHistoryStore.swift
//
//  NotificationSendLocalHistoryStore.swift
//  myfidpass
//
//  Historise côté app les envois de campagne afin d’alimenter Statistiques
//  immédiatement (HTTP 202) puis met à jour les compteurs après livraison confirmée.
//

import Foundation

struct NotificationSendLocalEntry: Codable, Sendable {
    var id: String
    var createdAtISO: String
    var title: String?
    var message: String
    var recipientOrSentCount: Int
    var expectedDevices: Int?
    var deliveryStatus: String?
    var jobId: String?
}

enum NotificationSendLocalHistoryStore {
    private static let storageKey = "myfidpass.notificationSendLocalHistory.v3"
    private static let perSlugMax = 40

    private static var pendingStatuses: Set<String> { ["queued", "sending", "pending"] }

    /// Clé de partition : compte connecté + commerce — évite tout mélange entre sessions ou commerces.
    private static func scopedStorageKey(slug: String) -> String {
        notificationHistoryScopedKey(slug: slug)
    }

    static func isPendingDeliveryStatus(_ raw: String?) -> Bool {
        pendingStatuses.contains(raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? "queued")
    }

    /// Enregistre une campagne **livrée** (ou état terminal) — jamais « Envoi en cours » dans l'historique.
    static func recordDelivered(
        slug: String,
        batchId: String,
        jobId: String?,
        title: String?,
        message: String,
        expectedDevices: Int,
        deliveryStatus: String,
        recipientsDistinct: Int
    ) {
        let scopeKey = scopedStorageKey(slug: slug)
        guard !scopeKey.isEmpty else { return }
        guard !isPendingDeliveryStatus(deliveryStatus) else { return }
        var map = readMap()
        var list = map[scopeKey] ?? []
        let iso = isoNow()
        let t = title?.trimmingCharacters(in: .whitespacesAndNewlines)
        let m = message.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !m.isEmpty else { return }
        let e = NotificationSendLocalEntry(
            id: batchId,
            createdAtISO: iso,
            title: (t?.isEmpty == false) ? t : nil,
            message: m,
            recipientOrSentCount: max(0, recipientsDistinct),
            expectedDevices: max(0, expectedDevices),
            deliveryStatus: deliveryStatus,
            jobId: jobId
        )
        if let idx = list.firstIndex(where: { $0.id == batchId }) {
            list[idx] = e
        } else if let last = list.last,
                  last.message == e.message,
                  last.title == e.title,
                  (isoDistanceSeconds(last.createdAtISO, e.createdAtISO) ?? 999) < 4 {
            return
        } else {
            list.append(e)
        }
        if list.count > perSlugMax { list = Array(list.suffix(perSlugMax)) }
        map[scopeKey] = list
        writeMap(map)
    }

    /// Suivi interne poll job → historique (mise à jour par batchId / jobId).
    static func updateDelivery(
        slug: String,
        batchId: String,
        jobId: String? = nil,
        deliveryStatus: String?,
        recipientsDistinct: Int,
        sentTotal: Int?
    ) {
        let status = deliveryStatus?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
        guard !isPendingDeliveryStatus(status) else { return }
        let scopeKey = scopedStorageKey(slug: slug)
        guard !scopeKey.isEmpty else { return }
        var map = readMap()
        guard var list = map[scopeKey] else { return }
        let job = jobId?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let idx = list.firstIndex(where: { entry in
            if !batchId.isEmpty, entry.id == batchId { return true }
            if let job, !job.isEmpty, entry.jobId == job { return true }
            return false
        }) else { return }
        var row = list[idx]
        if !batchId.isEmpty { row.id = batchId }
        row.deliveryStatus = deliveryStatus
        let delivered = max(recipientsDistinct, sentTotal ?? 0)
        if delivered > 0 { row.recipientOrSentCount = delivered }
        list[idx] = row
        map[scopeKey] = list
        writeMap(map)
    }

    static func entries(for slug: String) -> [NotificationSendLocalEntry] {
        let scopeKey = scopedStorageKey(slug: slug)
        guard !scopeKey.isEmpty else { return [] }
        return readMap()[scopeKey] ?? []
    }

    static func clearForAllSlugs() {
        UserDefaults.standard.removeObject(forKey: storageKey)
    }

    // MARK: - DTOs pour l’écran statistiques

    static func asCampaignInsights(_ entries: [NotificationSendLocalEntry]) -> [NotificationCampaignInsightDTO] {
        entries
            .filter { !isPendingDeliveryStatus($0.deliveryStatus) }
            .sorted { $0.createdAtISO > $1.createdAtISO }
            .map {
                let confirmed = max($0.recipientOrSentCount, 0)
                return NotificationCampaignInsightDTO(
                    batchId: $0.id,
                    triggerName: "manual",
                    createdAt: $0.createdAtISO,
                    sentTotal: confirmed > 0 ? confirmed : nil,
                    recipientsDistinct: confirmed,
                    returnedWithin48h: nil,
                    notificationTitle: $0.title,
                    message: $0.message,
                    sentPasskit: nil,
                    sentWebPush: nil,
                    deliveryStatus: $0.deliveryStatus,
                    expectedDevices: $0.expectedDevices
                )
            }
    }

    private static func readMap() -> [String: [NotificationSendLocalEntry]] {
        guard let d = UserDefaults.standard.data(forKey: storageKey) else { return [:] }
        return (try? JSONDecoder().decode([String: [NotificationSendLocalEntry]].self, from: d)) ?? [:]
    }

    private static func writeMap(_ m: [String: [NotificationSendLocalEntry]]) {
        if let d = try? JSONEncoder().encode(m) {
            UserDefaults.standard.set(d, forKey: storageKey)
        }
    }

    private static func isoNow() -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: Date())
    }

    private static func isoAgeSeconds(_ iso: String) -> TimeInterval? {
        isoDistanceSeconds(iso, isoNow())
    }

    private static func isoDistanceSeconds(_ a: String, _ b: String) -> TimeInterval? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let da = f.date(from: a), let db = f.date(from: b) { return abs(da.timeIntervalSince(db)) }
        f.formatOptions = [.withInternetDateTime]
        if let da = f.date(from: a), let db = f.date(from: b) { return abs(da.timeIntervalSince(db)) }
        return nil
    }
}

// MARK: - GET …/notifications/stats (cache disque / redémarrage)

/// Dernière réponse `campaigns` de l’endpoint stats notifs, par slug — évite une liste vide au cold start ou après échec réseau.
enum NotificationStatsEndpointCache {
    private static let key = "myfidpass.notificationStatsEndpointBySlug.v3"

    static func load(slug: String) -> [NotificationCampaignInsightDTO] {
        let scopeKey = notificationHistoryScopedKey(slug: slug)
        guard !scopeKey.isEmpty else { return [] }
        guard let d = UserDefaults.standard.data(forKey: key) else { return [] }
        guard let map = try? JSONDecoder().decode([String: [NotificationCampaignInsightDTO]].self, from: d) else { return [] }
        return (map[scopeKey] ?? []).filter { !NotificationSendLocalHistoryStore.isPendingDeliveryStatus($0.deliveryStatus) }
    }

    static func save(slug: String, campaigns: [NotificationCampaignInsightDTO]) {
        let scopeKey = notificationHistoryScopedKey(slug: slug)
        guard !scopeKey.isEmpty else { return }
        let deliveredOnly = campaigns.filter { !NotificationSendLocalHistoryStore.isPendingDeliveryStatus($0.deliveryStatus) }
        var map: [String: [NotificationCampaignInsightDTO]] = [:]
        if let d = UserDefaults.standard.data(forKey: key),
           let existing = try? JSONDecoder().decode([String: [NotificationCampaignInsightDTO]].self, from: d) {
            map = existing
        }
        map[scopeKey] = deliveredOnly
        if let out = try? JSONEncoder().encode(map) {
            UserDefaults.standard.set(out, forKey: key)
        }
    }

    static func clearAll() {
        UserDefaults.standard.removeObject(forKey: key)
    }
}

/// Clé UserDefaults : `userId|slug` (ou `email|slug`) — jamais slug seul (fuites multi-comptes).
private func notificationHistoryScopedKey(slug: String) -> String {
    let slugKey = slug.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    let account = AuthStorage.userId?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        ?? AuthStorage.userEmail?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        ?? ""
    guard !slugKey.isEmpty else { return "" }
    return account.isEmpty ? slugKey : "\(account)|\(slugKey)"
}

File: android/app/src/main/java/fr/myfidpass/data/local/NotificationSendLocalHistoryStore.kt
package fr.myfidpass.data.local

import android.content.Context
import fr.myfidpass.data.dto.NotificationCampaignInsightDto
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class NotificationSendLocalEntry(
    val id: String,
    val createdAtISO: String,
    val title: String? = null,
    val message: String,
    val recipientOrSentCount: Int = 0,
    val expectedDevices: Int? = null,
    val deliveryStatus: String? = null,
    val jobId: String? = null,
)

object NotificationSendLocalHistoryStore {
    private const val PREFS = "myfidpass.notificationSendLocalHistory.v3"
    private const val KEY = "by_scope"
    private const val PER_SLUG_MAX = 40
    private val json = Json { ignoreUnknownKeys = true }
    private val pendingStatuses = setOf("queued", "sending", "pending")

    fun isPendingDeliveryStatus(raw: String?): Boolean =
        pendingStatuses.contains(raw?.trim()?.lowercase().orEmpty().ifEmpty { "queued" })

    /** Enregistre une campagne livrée (état terminal) — jamais « Envoi en cours » dans l'historique. */
    fun recordDelivered(
        context: Context,
        slug: String,
        batchId: String,
        jobId: String?,
        title: String?,
        message: String,
        expectedDevices: Int,
        deliveryStatus: String,
        recipientsDistinct: Int,
    ) {
        val scopeKey = scopedStorageKey(context, slug)
        if (scopeKey.isEmpty() || message.isBlank()) return
        if (isPendingDeliveryStatus(deliveryStatus)) return
        val map = readMap(context).toMutableMap()
        val list = map[scopeKey]?.toMutableList() ?: mutableListOf()
        val iso = java.time.Instant.now().toString()
        val entry = NotificationSendLocalEntry(
            id = batchId,
            createdAtISO = iso,
            title = title?.trim()?.takeIf { it.isNotEmpty() },
            message = message.trim(),
            recipientOrSentCount = recipientsDistinct.coerceAtLeast(0),
            expectedDevices = expectedDevices.coerceAtLeast(0),
            deliveryStatus = deliveryStatus,
            jobId = jobId,
        )
        val idx = list.indexOfFirst { it.id == batchId || (jobId != null && it.jobId == jobId) }
        if (idx >= 0) {
            list[idx] = entry
        } else {
            val last = list.lastOrNull()
            if (last != null &&
                last.message == entry.message &&
                last.title == entry.title &&
                isoDistanceSeconds(last.createdAtISO, iso) < 4
            ) {
                return
            }
            list.add(entry)
        }
        map[scopeKey] = if (list.size > PER_SLUG_MAX) list.takeLast(PER_SLUG_MAX) else list
        writeMap(context, map)
    }

    fun entries(context: Context, slug: String): List<NotificationSendLocalEntry> =
        readMap(context)[scopedStorageKey(context, slug)].orEmpty()

    fun clearAll(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().clear().apply()
    }

    fun asCampaignInsights(entries: List<NotificationSendLocalEntry>): List<NotificationCampaignInsightDto> =
        entries
            .filter { !isPendingDeliveryStatus(it.deliveryStatus) }
            .sortedByDescending { it.createdAtISO }
            .map { e ->
                val confirmed = e.recipientOrSentCount.coerceAtLeast(0)
                NotificationCampaignInsightDto(
                    batchId = e.id,
                    triggerName = "manual",
                    createdAt = e.createdAtISO,
                    sentTotal = if (confirmed > 0) confirmed else null,
                    recipientsDistinct = confirmed,
                    returnedWithin48h = null,
                    notificationTitle = e.title,
                    message = e.message,
                    deliveryStatus = e.deliveryStatus,
                    expectedDevices = e.expectedDevices,
                )
            }

    private fun scopedStorageKey(context: Context, slug: String): String {
        val slugKey = slug.trim().lowercase()
        if (slugKey.isEmpty()) return ""
        val account = SessionStore(context).userEmail?.trim()?.lowercase().orEmpty()
        return if (account.isEmpty()) slugKey else "$account|$slugKey"
    }

    private fun isoDistanceSeconds(a: String, b: String): Long =
        runCatching {
            val da = java.time.Instant.parse(a)
            val db = java.time.Instant.parse(b)
            kotlin.math.abs(da.epochSecond - db.epochSecond)
        }.getOrDefault(999L)

    private fun readMap(context: Context): Map<String, List<NotificationSendLocalEntry>> {
        val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY, null)
            ?: return emptyMap()
        return runCatching {
            json.decodeFromString<Map<String, List<NotificationSendLocalEntry>>>(raw)
        }.getOrDefault(emptyMap())
    }

    private fun writeMap(context: Context, map: Map<String, List<NotificationSendLocalEntry>>) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY, json.encodeToString(map))
            .apply()
    }
}
